src/s_stuff.h

```diff
@@ -345,6 +356,57 @@ EXTERN void sys_rmpollfn(int fd);
 void sys_setalarm(int microsec);
 #endif
 
+/* tells whether Pd has capabilities for threaded I/O */
+int sys_hasthreadedio();
+
+/* Functions related to threaded I/O. */
+// register a socket for I/O
+// arguments:
+//   fd: the socket file descripter
+//   udp: is it a udp socket or not? (Used uniquely to preserve message boundaries or not)
+//   threaded: set to 1 to enable threaded I/O (if supported by Pd)
+//   user: user data passed to callback functions
+//   recvfn: a function which is called when the t_socket is ready for reading.
+//       sys_recv{from} should then be called on it
+//   senderrfn: a function which is called when threaded is true and a
+//       send{to}() fails. It must be provided when threaded is true and
+//       sys_send{to} will be called on this socket_t
+// returns:
+//    an opaque pointer to be used in sys_send, sys_sendto, sys_recv,
+//    sys_recvfrom and sys_bytes_available
+// The behaviour of the functions using the returned t_socket* is
+// feature-invariant, that is their behaviour depends uniquely on the arguments
+// passed here, regardless of whether Pd has threaded I/O support or not.
+EXTERN t_socket* sys_registersocket(int fd, int udp, int threaded, void *user, t_sockrecvfn recvfn, t_socksenderrfn senderrfn);
+
+// unregister socket. Returns the bare file descriptor
+EXTERN int sys_unregistersocket(t_socket *x);
+
+// drop-in replacements for recv{from} (internally fall back to recv/recvfrom if threaded is disabled)
+// if threaded is enabled, the flags argument is ignored
+EXTERN ssize_t sys_recv(t_socket *x, void* buf, size_t length, int flags);
```

Comment:
this obviously breaks building Pd with MSVC...

Reply:
It is worth nothing that `ssize_t` is POSIX, but not standard C.

IMO, the socket functions may just use `int` as the return value. I mean, nobody is ever going to send or receive buffers larger than 2 GB... (In fact, on Windows, `send` and `recv` still return `int`.)

Reply:
`d_soundfile.h` defines `ssize_t` https://github.com/pure-data/pure-data/blob/master/src/d_soundfile.h#L27-L39 , is that not a desirable approach?

Reply:
That's certainly a solution. Note that soundfiles might indeed be larger than 2 GB, so something like `ssize_t` is necessary. (Alternatively, it could have used `ptrdiff_t` (from `stdint.h`), or even Pd's own `t_int`.)

Reply:
I put ints in there for now, thanks all